FLINK-13871: Consolidate volatile status fields in StreamTask

pgaref opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

StreamTask has 3 volatile boolean fields (isRunning, canceled, failing) that are mutually exclusive and reflect different stages of stream task's lifecycle. We can represent them as a single enum state field.


## Brief change log

  - Introduced StreamTask enum TaskState and got rid of existing isRunning, canceled, failing volatile boolean fields


## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no** )
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

pgaref commented:
@flinkbot run azure

pgaref commented:
@flinkbot run azure

akalash commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -215,6 +215,9 @@
     private final StreamTaskActionExecutor actionExecutor;
 
+    /** Current state of the task, can be any of {@link TaskState}. */
+    private TaskState taskState;
```
Shouldn't it be volatile?

pgaref replied:
Yup, good catch Anton!

akalash commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -949,7 +952,9 @@ public final void cleanUp(Throwable throwable) throws Exception {
         getCompletionFuture().exceptionally(unused -> null).join();
         // clean up everything we initialized
-        isRunning = false;
+        if (!isCanceled() && !isFailing()) {
```
Technically, it is a race condition here. (as the same as several lines above in `failing` condition). We check the current state and only then set a new one. I don't think it is a big problem since we have finished the task already, but it is good to clarify this logic anyway.

Right now I see that we set the final state unconditionally in a couple of places while in a couple of other places, we check the condition that the state is not final yet. 

But I think actually that we have a pretty clear state machine here:
- init -> running
- running -> running(maybe)
- running -> canceled
- running -> failing
- running -> finished
So all other transfers should be ignored. So maybe we should come up with `compareAndSet` solution somehow?

@pnowojski , @rkhachatryan , WDYT?

P.S. Since the race condition is present in the current code as well we can implement it in the follow-up ticket rather than in this ticket.

pnowojski replied:
I would be cautious about any changes to the semantic here. Have you considered that in this PR you are flattening the matrix of valid combinations into the state machine that @akalash pointed out? For example before `cancelled` could be set `true` while `isRunning` was also `true`. Now that's no longer possible, and wouldn't such change in semantic cause some side-effects? 

Maybe it would be good to actually define the state machine that @akalash  proposed, but then also to go through all relevant getters (`isFailing()`/`isRunning()`/`isXXX()`), and check how are they being used and if they still make sense/are valid.

rkhachatryan replied:
I agree, FSM makes sense.

But I also agree, that all the combinations and changes are tricky.

So maybe if the purpose of the PR is to reduce `volatile` accesses, I'd rather move all three flags into the state object, and  reflect their exact same values there.
For example, replace `isRunning = false;` with `taskState = taskState.stopped();`  where `stopped()` returns a new state object with `running = false` inside.

On a side note, those `volatile` accesses aren't on the hot path so performance-wise it shouldn't change much.

pnowojski replied:
> On a side note, those volatile accesses aren't on the hot path so performance-wise it shouldn't change much

Yes, this doesn't affect performance. This ticket is about cleaning up the code and improving readability.

akalash replied:
I think that highly likely we can have FSM for this logic but I agree that it requires more research about its correctness.

So if we want to keep this ticket simple and save the logic that we have now. We can not have `failing` in the enum because we can have `failing` and `running` at the same time. Maybe we should have `TaskState` object with enum and `failing` flag inside. It doesn't look as clean as the solution in this PR  but at least it is cleaner than the current one and it keeps the same logic.

pgaref replied:
Thanks all for the comments, I have been thinking about a State machine all along, just didnt want to introduce too many changes on such a fundamental class. 

@pnowojski  `canceled` and `running` can never be true at the same time because of: https://github.com/apache/flink/pull/21923/files#diff-0c5fe245445b932fa83fdaf5c4802dbe671b73e8d76f39058c6eaaaffd9639faL977-L978

However, we can not make the same assumption for `failing` and `running` as  @akalash mentioned.

I have added the transitions to cover the Task lifecycle in the latest commit but I am leaning towards adding the failing flag as part of the TaskState to keep the logic closer to the existing.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -772,15 +773,15 @@ private CompletableFuture<Void> restoreGates() throws Exception {
     public final void invoke() throws Exception {
         // Allow invoking method 'invoke' without having to call 'restore' before it.
-        if (!isRunning) {
+        if (!isRunning()) {
```
wouldn't `if (taskState == INITIALIZED)` be the more correct equivalent here?

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -256,19 +259,27 @@
-    /** Flag to mark this task as canceled. */
-    private volatile boolean canceled;
+        private List<TaskState> validFromStates;
+        private boolean failing = false;
```
Why add the `failing` inside the `TaskState`?

akalash replied:
+1. It's better to avoid mutable fields in enums

akalash replied:
@1996fanrui , or if your question is why at all do we need to have `failing` as a field rather than as enum. it's because nobody wants to change the current logic in this PR. but currently `running` and `failing` can be `true` at the same time. (see this thread -https://github.com/apache/flink/pull/21923#discussion_r1111988132 )

1996fanrui replied:
After read the comment, I understand it. 

However, it isn't clear for other developers. I'm not sure it's the best way address the `failing` and `running`. Let me think more.

If so, I suggest add some comments in the code.

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -256,19 +259,27 @@
+            this.validFromStates = validFromStates;
+        }
+
+        public boolean isValidTransition(TaskState fromState, TaskState toState) {
```
The method isn't called, right?

As I understand, it should be called when transition state, right?

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -256,19 +259,27 @@
 
-    /** Flag to mark this task as canceled. */
-    private volatile boolean canceled;
+        private List<TaskState> validFromStates;
```
How about using `Set` instead of `List` due to we always call `validFromStates.contains()`?

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -256,19 +259,27 @@
-     * that early cancel() before invoke() behaves correctly.
+     * INITIALIZED: task constructor was called or on init state. RUNNING: task is in operation.
+     * FAILING: task is failing e.g., if an exception has occurred inside {@link #invoke()}.
+     * CANCELED: when this task is canceled. FINISHED: task successfully terminated.
```
How about drawing a state transition diagram? Similar to ExecutionState's annotation[1], it will be clearer.

[1] https://github.com/apache/flink/blob/2e91543836d667a0b367688bb5ce290c3164479c/flink-runtime/src/main/java/org/apache/flink/runtime/execution/ExecutionState.java#L26

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -256,19 +259,27 @@
     /**
-     * Flag to mark the task "in operation", in which case check needs to be initialized to true, so
-     * that early cancel() before invoke() behaves correctly.
+     * INITIALIZED: task constructor was called or on init state. RUNNING: task is in operation.
```
How about moving the description of the state to the corresponding state field? It's similar to `JobStatus`[1] or `ExecutionState`[2].

[1] https://github.com/apache/flink/blob/2e91543836d667a0b367688bb5ce290c3164479c/flink-core/src/main/java/org/apache/flink/api/common/JobStatus.java#L26
[2] https://github.com/apache/flink/blob/2e91543836d667a0b367688bb5ce290c3164479c/flink-runtime/src/main/java/org/apache/flink/runtime/execution/ExecutionState.java#L57

1996fanrui reviewed (commented):
Hi @pgaref , thanks a lot for the refactor, it's very useful for maintainability. I have left some comments, please help take a look in your free time, thanks~

Thanks @pnowojski @akalash @rkhachatryan 's review. I found FLINK-29816[1] related to this TaskState, and I think it should be fixed first, what do you think?

[1] https://issues.apache.org/jira/browse/FLINK-29816?focusedCommentId=17692203&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17692203

pgaref commented:
> Hi @pgaref , thanks a lot for the refactor, it's very useful for maintainability. I have left some comments, please help take a look in your free time, thanks~
> 
> Thanks @pnowojski @akalash @rkhachatryan 's review. I found [FLINK-29816](https://issues.apache.org/jira/browse/FLINK-29816)[1] related to this TaskState, and I think it should be fixed first, what do you think?
> 
> [1] https://issues.apache.org/jira/browse/FLINK-29816?focusedCommentId=17692203&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17692203

@1996fanrui thanks for the comments!
Let me rework the patch addressing the comments now that [FLINK-29816](https://issues.apache.org/jira/browse/FLINK-29816) is merged :)

pgaref commented:
Thanks everyone for the useful comments! PTAL on the latest changes, in short:

- Checked all state transitions and documented them as part of https://github.com/apache/flink/pull/21923/files#diff-0c5fe245445b932fa83fdaf5c4802dbe671b73e8d76f39058c6eaaaffd9639faR268
- Integrated **Restoring** state introduced as part of [FLINK-29816](https://issues.apache.org/jira/browse/FLINK-29816)
- Added **Failing** as a separate State boolean field in TaskSate object (can always rework this as a followup)
- Refrained from adding state transition validation as this is already happening as part of Task ExecutionState https://github.com/apache/flink/blob/8be94e6663d8ac6e3d74bf4cd5f540cc96c8289e/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L907

Thoughts?

akalash commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -1019,16 +1049,24 @@ public CanEmitBatchOfRecordsChecker getCanEmitBatchOfRecords() {
         return () -> !this.mailboxProcessor.hasMail() && taskIsAvailable();
     }
 
+    public final boolean isInitialized() {
```
I'm not sure that we need these new methods(especially as public) considering that right now nobody uses them. Maybe it makes sense to create some method in the TaskState like `taskState.isStatus(RUNNING)`. But at least we should avoid public methods which we don't need.

rkhachatryan replied:
Or it could be something like
```
private static class TaskState {
    ...
    public final boolean isRestoring() {
        return this.status == TaskState.Status.RESTORING;
    }
}

pgaref replied:
Ack, moved to TaskState class

akalash commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -255,23 +258,52 @@
-    /** Flag to mark this task as canceled. */
-    private volatile boolean canceled;
+    /** Possible states of a Task. */
+    private static class TaskState {
```
Maybe is it better to put the inner class at the bottom of this class?

pgaref replied:
Ack. moved to the bottom of the class

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -733,8 +764,7 @@ void restoreInternal() throws Exception {
 
-        isRunning = true;
-        isRestoring = false;
+        taskState.status = TaskState.Status.RUNNING;
```
At this point, the task might be already cancelled, right?

Previously, this was captured by `isRunning = true` and `isCancelled = true`.
But now, we overwrite `CANCELLED` by `RUNNING`.

This might break checks like [this one](https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java#L339) in `SourceStreamTask` and cause task and then job failure.

---

I guess this is basically the same concern expressed previously that we might lose some transitions.
I see only two ways to be on the safe side:
1. Have all the same flags, but inside the inner `TaskState` class
1. Model the enum exactly as `2^number_of_flags` AND transition state via CAS (not just assignment)

WDYT?

pgaref replied:
Hey Roman, very good point since TaskCanceler is asynchronous in nature.
However, the only way to invoke [Task cancelation](https://github.com/apache/flink/blob/8be94e6663d8ac6e3d74bf4cd5f540cc96c8289e/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L1662) is through the [invokable method](https://github.com/apache/flink/pull/21923/files#diff-0c5fe245445b932fa83fdaf5c4802dbe671b73e8d76f39058c6eaaaffd9639faL982-L983) that simultaneously marks the task as **not** Running and Canceled. 

In that sense, its not the same situation as failing, so I believe we are safe here.

rkhachatryan replied:
Hey Panos,
> the [invokable method](https://github.com/apache/flink/pull/21923/files#diff-0c5fe245445b932fa83fdaf5c4802dbe671b73e8d76f39058c6eaaaffd9639faL982-L983) that simultaneously marks the task as not Running and Canceled

Yes, but after that, the Task Thread might update `isRunning`, but not `cancelled`.

1. Task Thread enters `restoreInternal` and passes `ensureNotCanceled`
2. Canceller Thread calls `invokable.cancel` and sets `isRunning = false`, `canceled = true`
3. Task Thread continues and sets `isRunning = true`, `isRestoring = false`
4. Legacy Source Thread gets interrupted and checks `isCancelled` - it's `true`, so `CancelTaskException` is thrown, which won't fail the job

With this PR, all the flags are combined and updated unconditionally:
1. Task Thread enters `restoreInternal` and passes `ensureNotCanceled`
2. Canceller Thread calls `invokable.cancel` and sets `taskState.status = TaskState.Status.CANCELED`
3. Task Thread continues and sets `taskState.status = TaskState.Status.RUNNING`
4. Legacy Source Thread gets interrupted and checks `isCancelled`, i.e. `taskState.status == TaskState.Status.CANCELED`; which is `false`, so some other `Throwable` is thrown, which **does** fail the job

akalash replied:
As I understand right now all transitions are correct. The only problem is we can overwrite the final status with another final status which is incorrect behaviour. To fix that, I agree we can use CAS to update TaskStateStatus and allow only transitions described in javadoc.

I also don't think that we should get back to 4 flags since we already spend a lot of time understanding which combinations of them are really possible and it seems nobody knows for sure. So it is better to decrease these combinations and I think the enum with clear transitions is good choice.

I still think that we should even get rid of `failing` flag since I feel we can combine it with enum but I'm ok if we do it later

rkhachatryan replied:
> now all transitions are correct

I'd add CANCELLING state to reflect the above combination of flags.

> I also don't think that we should get back to 4 flags since we already spend a lot of time understanding which combinations of them are really possible

Makes sense. I'm not insisting on any option. As for the knowledge, should we encode it as allowed state transitions? I believe it was there previously in the PR, but then was deleted. 

> I still think that we should even get rid of failing

👍 and it seems just necessary if we proceed with CAS

akalash replied:
Potentially we can have CANCELLING status but right now I don't see where we can actually use it. For example, if we consider the above-described scenario, we have a situation when canceled=true at the same time as isRestoring=true as well. isRestoring we use only inside `handleAsyncException` and if we want to keep this logic we indeed need some extra status like CANCELLING but I actually think that we don't need to keep that logic since if the job is canceled already we can easily ignore other exceptions which can happen since they can be actually because of cancelation. I have the same thoughts about canceled=true + isRunning=true.(I don't see where we should execute logic isRunning=true if the canceled is true already.)
Long story short, if we indeed understand where we need CANCELLING status let's add it but if we need it only for keeping the old behavior let's discuss this old behavior since I am not really sure that the old behavior is correct.

rkhachatryan replied:
> we don't need to keep that logic since if the job is canceled already we can easily ignore other exceptions which can happen since they can be actually because of cancelation

Please note that Task cancellation is not the same as Job cancellation. For example, in case of some intermittent failure, the Job will be `RESTARTING`, but the tasks will be cancelled and then rescheduled. Ideally, we shouldn't be ignoring **any** exception during restart (just interruptions and cancellations). Otherwise, it might create a resource leak (job correctness won't be affected though).

As for the `running` flag (which will report `true` on `master`; and `false` when CAS to `CANCELLED`);
I briefly skimmed though its usages:
- `Task.restoreAndInvoke` -> `StreamTask.invoke` -> `StreamTask.restoreInternal` - regression: it will allow 2nd `restoreInternal`
- ongoing checkpoints will not be aborted by RPC - that's fine
- local state will not be pruned by CHK abort RPC - regression (but **likely** will be pruned on the next checkpoint)
- ...

The above can probably be fixed (in this PR), but my point is that it's hard to check all the combinations and their usages.
If that's true, I'd stick closer to the original behavior by having states **unless** we are sure that they are impossible.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -980,8 +963,7 @@ protected CompletableFuture<Void> getCompletionFuture() {
     public final void cancel() throws Exception {
-        isRunning = false;
-        canceled = true;
+        taskState.canceled = true;
```
Shouldn't we also update `taskState.status` here?

(ideally, in my view we should use either enum or a set of flags inside the `TaskState`, but not both)

rkhachatryan replied:
> maintain the existing logic where both flags (running and canceled are true)

I see. Have you considered adding `CANCELLING` (in addition to `CANCELLED`) constant?

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -1797,4 +1779,74 @@ public interface CanEmitBatchOfRecordsChecker {
+         *                 |             |           |
+         *                 |             |           |
+         *                 |             V           |
+         *                 |           CANCELED -----+
```
`CANCELED` state was removed from the code.

dmvk commented:
@pgaref @rkhachatryan @akalash, what's the next step for this effort?
